File: apps/worker/src/services/attack-chain-builder.ts
// Copyright (C) 2025 Keygraph, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License version 3
// as published by the Free Software Foundation.

/**
 * Attack chain builder
 *
 * Assembles multi-step attack chains from shared knowledge accumulated
 * by prior agents (framework analysis, endpoint inventory, frontend routes,
 * vulnerability findings).
 */

import type { ActivityLogger } from '../types/activity-logger.js';
import type { SharedKnowledge } from '../types/shared-knowledge.js';
import type { AttackChain } from './route-chain-builder.js';
import { buildAttackChainsFromAnalysis } from './route-chain-builder.js';

export type { AttackChain, AttackChainStep } from './route-chain-builder.js';

/**
 * Build attack chains from shared knowledge.
 * Combines framework analysis, endpoint inventory, frontend routes,
 * and vulnerability findings into complete attack scenarios.
 */
export async function buildAttackChains(
  knowledge: SharedKnowledge,
  logger: ActivityLogger,
): Promise<AttackChain[]> {
  const frameworkEndpoints = knowledge.frameworkAnalysis?.inferredEndpoints ?? [];
  const frontendRoutes = knowledge.frontendRoutes?.routes ?? [];
  const xssChains = knowledge.frontendRoutes?.xssVectors ?? [];

  // 1. Build chains from framework + frontend correlation
  const analysisChains = buildAttackChainsFromAnalysis(frameworkEndpoints, frontendRoutes, xssChains, logger);

  // 2. Enhance chains with vulnerability context if available
  const vulnKnowledge = knowledge.vulnerabilityContext;
  if (vulnKnowledge) {
    for (const chain of analysisChains) {
      // Check if any endpoint in this chain has confirmed vulnerabilities
      for (const step of chain.steps) {
        const vulnEntries = vulnKnowledge.endpointVulnerabilities[step.endpoint];
        if (vulnEntries && vulnEntries.length > 0) {
          const confirmed = vulnEntries.some((v) => v.confirmed);
          if (confirmed && chain.confidence !== 'confirmed') {
            // Upgrade confidence — mutation of the chain object is safe here
            // because these are freshly created in this function
            (chain as { confidence: string }).confidence = 'confirmed';
          }
        }
      }
    }
  }

  logger.info(`Built ${analysisChains.length} attack chain(s) from shared knowledge`);
  return analysisChains;
}

File: packages/core/src/shannon_core/services/attack_chain_builder.py
"""Attack chain builder.

Assembles multi-step attack chains from framework analysis and frontend
mapping results.
Ported from /root/shannon/apps/worker/src/services/attack-chain-builder.ts
"""

from __future__ import annotations

import logging

from shannon_core.services.framework_analyzer import FrameworkAnalysisResult
from shannon_core.services.frontend_mapper import FrontendAnalysisResult
from shannon_core.services.route_chain_builder import (
    AttackChain,
    build_attack_chains_from_analysis,
)


async def build_attack_chains(
    framework_result: FrameworkAnalysisResult,
    frontend_result: FrontendAnalysisResult,
    logger: logging.Logger,
) -> list[AttackChain]:
    """Build attack chains from analysis results.

    1. Call build_attack_chains_from_analysis() for base chains
    2. Return complete attack chain list
    """
    framework_endpoints = framework_result.inferred_endpoints
    frontend_routes = frontend_result.routes
    xss_chains = frontend_result.xss_chains

    chains = build_attack_chains_from_analysis(
        framework_endpoints, frontend_routes, xss_chains, logger,
    )

    logger.info("Built %d attack chain(s) from shared knowledge", len(chains))
    return chains
